docs/design/validation.md

```diff
@@ -0,0 +1,1989 @@
+
+For example:
+
+```go
+&schema.Resource{
+    // ...
+    CustomizeDiff: func(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
+        if value := diff.Get("attribute_one").(string); value == "special condition" {
+            if _, ok := diff.GetOk("attribute_two"); !ok {
+                return fmt.Errorf("'attribute_two' must be set when 'attribute_one' is %q", value)
+            }
+        }
+
+        return nil
+    },
+}
+```
+
+In general, `CustomizeDiff` is not broadly utilized across the ecosystem due to the complexity of properly implementing and testing the functionality.
+
+##### `helper/customdiff` Package
+
+Similar to how the `helper/validation` package of common functionality was created for `ValidateFunc`, a [`helper/customizediff`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff) package was created for common `CustomizeDiff` functionality.
+
+In terms of validation, this provided helpers such as:
+
+```go
+&schema.Resource{
+    // ...
+    CustomizeDiff: customdiff.IfValue(
+        "attribute_one",
+        func(ctx context.Context, rawValue, meta interface{}) bool {
+            value, ok := rawValue.(string)
+
+            // potentially difficult to diagnose type issue
+            if !ok {
+                return false
+            }
+
+            return value == "special condition"
+        },
+        customdiff.ValidateValue(
+            "attribute_two",
+            func(ctx context.Context, rawValue, meta interface{}) error {
+                value, ok := rawValue.(string)
+
+                if !ok {
+                    return fmt.Errorf("incorrect type conversion for attribute_two")
+                }
+
+                if value != "" {
+                    return fmt.Errorf("cannot provide attribute_two value when attribute_one is \"special condition\"")
+                }
+
+                return nil
+            },
+        ),
+    ),
+}
+```
+
+These likely would have been simplified into further helpers should there have been more `CustomizeDiff` usage.
+
+## Goals
+
+This framework design should strive to accomplish the following with validation support.
+
+Allow provider developers access to all current types of provider validation:
+
+- Provider configurations
+- Resource configurations (and configurations during plans)
+- Data Source configurations (and configurations during plans)
+
+Including where possible:
+
+- Single attribute value validation (e.g. string length)
+- Multiple attribute validation (e.g. attributes or attribute values that conflict with each other)
+
+In terms of implementation, the following core concepts:
+
+- Low level primitives (e.g other portions of the framework, external packages, and provider developers can implement higher level functionality)
```

Comment:
This sounds a bit like plugin-go to me. Is it fair to frame this instead as "composable building blocks"?

Reply:
Indeed! Will be fixed in the next commit. Thanks for calling this out.